### ISSUE TITLE
update-component-sha.sh doesn't work on bash shell(?)

### ISSUE THREAD

**Issue by arm64b** [linuxkit/linuxkit#2284]
**Description**
Under ubuntu 16.04 using /bin/bash on both amd64 and arm64:

root@entos-huawei-d05:/home/mainline/scripts# ./update-component-sha.sh 
./update-component-sha.sh: 10: ./update-component-sha.sh: Syntax error: "(" unexpected

Change the first line to /bin/bash will be OK. Trivial issue, just as a remind.

**Steps to reproduce the issue:**

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by arm64b** [comment]
Ah, today I am trying to clone the LinuxKit mainline in my ARM64 server, first I need to build the alpine image with` $make ORG=arm64b`  after that I get an arm64 alpine docker image-- arm64b/alpine:c3149ca5db79f2bcd9525491f783fc835d4845cf

Then I want to replace all the 'FROM linuxkit/alpine: xxx' with 'FROM arm64b/alpine:xxx', so I run the 'update-component-sha.sh' like this:
mainline/scripts$./update-component-sha.sh  --image arm64b/alpine c3149ca5db79f2bcd9525491f783fc835d4845cf
sed: no input files

Clearly we need a useful tool to reach the goal like I did in my case. For the tool itself, am I missing something?

**Comment by deitch** [comment]
The script is written for `sh`, works fine in `bash`.

In Ubuntu, looks like the default shell is `dash`:

```
# ls -l /bin/sh
lrwxrwxrwx 1 root root 4 Feb 17  2016 /bin/sh -> dash
```

I don't know how `dash`'s semantics differ from `sh`, but clearly enough that it is unhappy.

**Comment by deitch** [comment]
> Then I want to replace all the 'FROM linuxkit/alpine: xxx' with 'FROM arm64b/alpine:xxx',

This may be a valid use case, but the tool itself is made to replace `sha` tags _after_ the image name, not the image name _itself_.

> mainline/scripts$./update-component-sha.sh  --image arm64b/alpine c3149ca5db79f2bcd9525491f783fc835d4845cf

The above will replace the tag on all instances of `arm64b/alpine` to `c3149...`. It will *not* replace the image name from `linuxkit/alpine` to `arm64b/alpine`.

**Comment by arm64b** [comment]
Seems ubuntu use bash as the default shell:
`#env
...
SHELL=/bin/bash
...`

But my point is how to replace 'linuxkit/alpine:xxx' to 'arm64b/alpine:xxx'...

**Comment by arm64b** [comment]
Since the build process give the chance for a user to override the 'ORG' variable, so I think it will be very useful to replace 'linuxkit/alpine:sha' to 'arm64b/alpine:sha' such kind of usage case, else I will replace all the linuxkit/alpine manually.

**Comment by deitch** [comment]
I didn't boot up my Ubuntu machine, just run `docker run -it --rm ubuntu:16.04 sh`, but it definitely symlinked `sh -> dash`.

> But my point is how to replace 'linuxkit/alpine:xxx' to 'arm64b/alpine:xxx'...

Open a PR for a tool to do it? I would get feedback on people as to whether they want to extend the existing script or add a new one before putting the effort it. Tag with the people who were active on the update script discussion yesterday at #2275

**Comment by deitch** [comment]
> Since the build process give the chance for a user to override the 'ORG' variable, so I think it will be very useful to replace 'linuxkit/alpine:sha' to 'arm64b/alpine:sha' such kind of usage case, else I will replace all the linuxkit/alpine manually

Oh, you want to replace it during build? Doesn't the makefile already support overriding both `ORG` and `HASH`? I know it does for anything in `pkg/`. Look in `pkg/package.mk` and extend the `tools/alpine/Makefile` to support a similar var-based override?

**Comment by arm64b** [comment]
> Oh, you want to replace it during build?

No, I mean I will override the ORG during the build like this:
`$make ORG=arm64b`

After that, I want to use update-component-sha.sh to replace the linuxkit/alpine:sha with arm64b/alpine:sha under the folder 'pkg' and 'tools'.

**Comment by deitch** [comment]
> After that, I want to use update-component-sha.sh to replace the linuxkit/alpine:sha with arm64b/alpine:sha under the folder 'pkg' and 'tools'.

Ah. Then back to the original. You might want to create a PR that either updates the existing script or creates a new one.

**Comment by deitch** [comment]
I do wonder if that is the best way, though. Probably worth a bit of effort to make all of the `pkg/` sections work natively with arm64, i.e. without having to replace the sha every time you build.

Side note: you work with any of the Packet.net team?

**Comment by rn** [comment]
@arm64b there is a proposal to make this easier with the `moby` tool https://github.com/moby/tool/issues/95

Also, I now have a arm64 machine available and will try to work on making `linuxkit/alpine` a multiarch package. With that it should be easier to also turn the other packages into multiarch packages. I should probably create a tracking issue with some details of what I have in mind...

**Comment by arm64b** [comment]
@deitch I don't work with Packet.net team
@rn That's great, so I think it will not be very long before we can announce the official arm64 support for LinuxKit project 👍

**Comment by ijc** [comment]
> the tool itself is made to replace sha tags after the image name, not the image name itself.

As I've [pointed out before](https://github.com/linuxkit/linuxkit/pull/2275#discussion_r128997311), this isn't how I intended it when I originally wrote this script nor is it how I've been using it until now. I use it as:
```
./scripts/update-component-sha.sh  linuxkit/getty:08b704915af0ce90f8f40df5d41d4c1aa14ef83a  linuxkit/getty:97c5e2c8ebad23c2ed743366b475b5c15c42f70e
```
I understood that #2275 wasn't breaking that use case in practice (even if the docs pointed towards another way of using it) and I want it to continue to be supported going forward.

Given that IMHO this should work (or it should be fixed to work):
```
./scripts/update-component-sha.sh linuxkit/alpine:f0bec4717a4b3f9d068d1b72626f3bf938d7365a arm64b/alpine:f0bec4717a4b3f9d068d1b72626f3bf938d7365a
```

**Comment by ijc** [comment]
WRT the default shell on Ubuntu, the setting of `$SHELL` in ones environment is irrelevant, only the `#!` line in the script matters and that is `/bin/sh` which by default is a symlink to `/bin/dash` on Ubuntu and Debian. It can be switched to be `/bin/bash` (but note that when invoked as `/bin/sh` Bash enters a POSIX compatibility mode. Note that this is also independent to the default login shell, which typically defaults to `/bin/bash` and is specified per user in `/etc/passwd`.

For those who don't know `dash` is intended as a small/fast POSIX shell. It's slightly more pragmatic than `posh` (which is a strictly by the books POSIX shell) but the number of extensions are small. In the context of Debian (and by extension Ubuntu) dash satisfies [policy 10.4](https://www.debian.org/doc/debian-policy/ch-files.html#s10.4) (I think quite precisely, i.e. no more and no less) which requires SuSv3 + a small set of extensions listed there.

**Comment by ijc** [comment]
Seems like a bashism has crept into the script:
```
$ checkbashisms scripts/update-component-sha.sh
possible bashism in scripts/update-component-sha.sh line 10 ('function' is useless):
function usage() {
```

**Comment by deitch** [comment]
> As I've pointed out before, this isn't how I intended it when I originally wrote this script nor is it how I've been using it until now. I use it as

Oh right. Well, if it works in `script.sh some/image:hash otherorg/funnyimage:diffhash` then all good, and if not, should be updated.

**Comment by ijc** [comment]
From reading the code I think `script.sh some/image:hash otherorg/funnyimage:diffhash` should indeed work once the bashism is resolved.

**Comment by deitch** [comment]
> should indeed work once the bashism is resolved.

Checking now...

**Comment by deitch** [comment]
I can confirm it does. Nice catch on the bashism. PR very shortly...

**Comment by deitch** [comment]
PR in #2285

**Comment by deitch** [comment]
Closed via #2285
